src/main/kotlin/io/inventi/eventstore/util/script/CopyAndReplaceUtils.kt

```diff
@@ -92,74 +95,83 @@ data class CopyAndReplaceOperation(
+                EventData.newBuilder()
+                        .type(event.eventType)
+                        .jsonData(it)
+                        .jsonMetadata(meta)
+                        .build()
+            }
 
-    private fun transformEvents(streamId: String, eventNumbers: List<Long>, transformFn: (event: RecordedEvent, mapper: ObjectMapper) -> ByteArray?): TransformedEventsData {
-        val firstEventNumber = this.eventStore.getStreamMetadata(streamId)
-                .thenApply { it.streamMetadata.truncateBefore }
-                .get()
-                ?: 0
-
-        var initialExpectedVersion: Long = firstEventNumber
-        val eventData = this.eventStore.iterateStreamEventsForward(streamId, firstEventNumber, bufferSize, true)
-                .asSequence().mapNotNull { resolvedEvent ->
-                    val event = resolvedEvent.event
-                    initialExpectedVersion = event.eventNumber
-                    val (data, meta) = if (resolvedEvent.originalEventNumber() in eventNumbers) {
-                        val newData = transformFn(event, objectMapper)
-                        newData to transformMetadata(event, objectMapper, shouldPutId = true)
-                    } else {
-                        event.data to transformMetadata(event, objectMapper, shouldPutId = true)
-                    }
-
-                    println("#${event.eventNumber}: ${meta?.toString(Charsets.UTF_8)} --> ${data?.toString(Charsets.UTF_8)}")
-                    data?.let {
-                        event to EventData.newBuilder()
-                                .type(event.eventType)
-                                .jsonData(data)
-                                .jsonMetadata(meta)
-                                .build()
-                    }
-
-                }.toMap()
-        return TransformedEventsData(initialExpectedVersion, eventData)
+        }
+        appendEvents(streamId, events, expectedStreamVersion)
+        truncateBefore(expectedStreamVersion + 1, streamId)
     }
 
-    private fun appendEvents(streamId: String, events: Collection<EventData>, initialExpectedVersion: Long) {
-        eventStore.startTransaction(streamId,  initialExpectedVersion).get().use { transaction ->
-            events.chunked(writeChunkSize).forEach { events ->
-                transaction.write(events).get()
-            }
-            transaction.commit().get()
-        }
+    fun copyAndReplace(streamId: String, transformFn: (event: ResolvedEvent, mapper: ObjectMapper) -> EventData?) {
+        val (expectedStreamVersion, events) = transformEvents(streamId, transformFn)
+        appendEvents(streamId, events, expectedStreamVersion)
+        truncateBefore(expectedStreamVersion + 1, streamId)
     }
 
-    private fun transformMetadata(event: RecordedEvent, objectMapper: ObjectMapper, shouldPutId: Boolean): ByteArray? {
+    fun copyMetadata(event: RecordedEvent, shouldPutOverrideIds: Boolean = true): ByteArray? {
```

Comment:
generally it is bad idea to pass boolean as parameter
In case we will use `shouldPutOverrideIds` somewhere, we should have separate function IMO